README.md
<!--
This README describes the package. If you publish this package to pub.dev,
this README's contents appear on the landing page for your package.

For information about how to write a good package README, see the guide for
[writing package pages](https://dart.dev/guides/libraries/writing-package-pages).

For general information about developing packages, see the Dart guide for
[creating packages](https://dart.dev/guides/libraries/create-library-packages)
and the Flutter guide for
[developing packages and plugins](https://flutter.dev/developing-packages).
-->

A package that provides a sheet widget that can be drag to other positions

## Demo
<p align='center'>
    <img src="https://raw.githubusercontent.com/XuanTung95/comment_sheet/master/example/demo.mp4" width = "280" />
</p>

## Usage

Please check `example` folder for usage.



CHANGELOG.md
## 0.0.1 - (2023-Jan-05)

* First release


test/comment_sheet_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:comment_sheet/comment_sheet.dart';

void main() {

}


lib/src/comment_sheet.dart
import 'dart:math' as math;
import 'package:flutter/foundation.dart';
import 'package:flutter/gestures.dart';
import 'package:flutter/material.dart';
import 'package:flutter/physics.dart';
import 'package:flutter/widgets.dart';

import 'comment_sheet_controller.dart';

enum WidgetPosition { below, above }

class CommentSheet extends StatefulWidget {
  const CommentSheet({
    Key? key,
    required this.child,
    required this.grabbing,
    required this.slivers,
    required this.calculateTopPosition,
    required this.commentSheetController,
    this.initTopPosition = 0,
    this.scrollController,
    this.topWidget,
    this.bottomWidget,
    this.topPosition = WidgetPosition.below,
    this.grabbingPosition = WidgetPosition.below,
    this.onPointerUp,
    this.onPointerDown,
    this.onPointerCancel,
    this.backgroundBuilder,
    this.simulationBuilder = CommentSheetState.buildSimulation,
    this.scrollPhysics = const CommentSheetBouncingScrollPhysics(
        parent: AlwaysScrollableScrollPhysics()),
    this.onAnimationComplete,
    this.clipBehavior = Clip.hardEdge,
    this.onTopChanged,
  }) : super(key: key);

  final Widget? child;
  final Widget? grabbing;
  final Widget Function(CommentSheetInfo info)? topWidget;
  final double initTopPosition;
  final WidgetPosition topPosition;
  final Widget? bottomWidget;
  final WidgetPosition grabbingPosition;
  final List<Widget> slivers;
  final ScrollController? scrollController;
  final CommentSheetController commentSheetController;
  final WidgetBuilder? backgroundBuilder;
  final double Function(CommentSheetInfo info) calculateTopPosition;

  final void Function(BuildContext context, CommentSheetInfo info)? onPointerUp;
  final void Function(BuildContext context, CommentSheetInfo info)? onPointerCancel;
  final void Function()? onPointerDown;

  final void Function(BuildContext state, CommentSheetInfo info)?
      onAnimationComplete;

  final ScrollPhysics scrollPhysics;

  final Simulation Function(double target, CommentSheetInfo info)
      simulationBuilder;

  final Clip clipBehavior;

  final void Function(double top)? onTopChanged;

  @override
  State<CommentSheet> createState() => CommentSheetState();
}

class CommentSheetInfo {
  final BoxConstraints size;
  final VelocityTracker velocity;
  final double currentTop;
  final ScrollController scrollController;
  final double animatingTarget;
  final bool isAnimating;

  const CommentSheetInfo({
    required this.size,
    required this.velocity,
    required this.currentTop,
    required this.scrollController,
    required this.animatingTarget,
    required this.isAnimating,
  });
}

class CommentSheetState extends State<CommentSheet>
    with SingleTickerProviderStateMixin {
  late ScrollController scrollController;
  late CommentSheetController commentSheetController;
  late AnimationController animationController;
  late final bool _createdScrollController;

  double _top = 0; // from top of stack -> top of the grabbing
  double _scrollOffset = 0; // scrollController.offset when offset < 0

  BoxConstraints _size = BoxConstraints.tight(Size.zero);

  BoxConstraints get size => _size;

  final VelocityTracker _vt = VelocityTracker.withKind(PointerDeviceKind.touch);

  VelocityTracker get velocityTracker => _vt;

  double get fakeTop => _top - _scrollOffset;
  bool _movingGrabbing = false;
  double _animatingTarget = 0;

  @override
  void initState() {
    super.initState();
    _setCommentSheetController();
    _top = widget.initTopPosition;
    scrollController = widget.scrollController ?? ScrollController();
    _createdScrollController = widget.scrollController == null;
    scrollController.addListener(_scrollControllerListener);

    animationController = AnimationController.unbounded(vsync: this);
    animationController.addListener(() {
      setState(() {
        _top = animationController.value;
        widget.onTopChanged?.call(fakeTop);
      });
    });
  }

  void _setCommentSheetController() {
    commentSheetController = widget.commentSheetController;
    commentSheetController.state = this;
  }

  void _scrollControllerListener() {
    if (scrollController.offset <= 0) {
      setState(() {
        _scrollOffset = scrollController.offset;
        widget.onTopChanged?.call(fakeTop);
      });
    } else {
      // scrollOffset should be 0
      if (_scrollOffset != 0) {
        setState(() {
          _scrollOffset = 0;
          widget.onTopChanged?.call(fakeTop);
        });
      }
    }
  }

  @override
  void didUpdateWidget(covariant CommentSheet oldWidget) {
    super.didUpdateWidget(oldWidget);
    _setCommentSheetController();
  }

  @override
  void dispose() {
    scrollController.removeListener(_scrollControllerListener);
    commentSheetController.state = null;
    animationController.dispose();
    if (_createdScrollController) {
      scrollController.dispose();
    }
    super.dispose();
  }

  TickerFuture animateToPosition(double target) {
    _animatingTarget = target;
    var simulation = widget.simulationBuilder.call(target, getInfo(_size));
    final ret = animationController.animateWith(simulation);
    if (widget.onAnimationComplete != null) {
      ret.whenComplete(() {
        widget.onAnimationComplete?.call(context, getInfo(_size));
      });
    }
    return ret;
  }

  static Simulation buildSimulation(double target, CommentSheetInfo info) {
    return StopBouncingScrollSimulation(
      spring: SpringDescription.withDampingRatio(
        mass: 0.5,
        stiffness: 200.0,
        ratio: 1.1,
      ),
      position: info.currentTop,
      velocity: 0,
      // velocity: info.velocity.getVelocity().pixelsPerSecond.dy,
      leadingExtent: target,
      trailingExtent: target,
      tolerance: Tolerance.defaultTolerance,
    );
  }

  void _resetTopToCurrentScrollOffset() {
    if (_scrollOffset < 0) {
      _top = _top - _scrollOffset;
      scrollController.jumpTo(0);
      setState(() {});
    }
  }

  @override
  Widget build(BuildContext context) {
    return LayoutBuilder(builder: (context, size) {
      _size = size;
      final info = getInfo(size);
      Widget? widgetTop =
          widget.topWidget == null ? null : widget.topWidget!.call(info);

      return Stack(
        clipBehavior: widget.clipBehavior,
        children: [
          if (widget.child != null) RepaintBoundary(
            child: widget.child!
          ),
          if (widget.topPosition == WidgetPosition.below && widgetTop != null)
            widgetTop,
          if (widget.backgroundBuilder != null)
            Positioned(
              top: fakeTop,
              left: 0,
              right: 0,
              bottom: 0,
              child: widget.backgroundBuilder!.call(context),
            ),
          Positioned(
            top: _top,
            left: 0,
            right: 0,
            bottom: 0,
            child: Listener(
              onPointerDown: (PointerDownEvent p) {
                _vt.addPosition(p.timeStamp, p.position);
                widget.onPointerDown?.call();
              },
              onPointerMove: (detail) {
                _vt.addPosition(detail.timeStamp, detail.position);
              },
              onPointerCancel: (detail) {
                _resetTopToCurrentScrollOffset();
                widget.onPointerCancel?.call(
                  context,
                  getInfo(size),
                );
              },
              onPointerUp: (detail) {
                _resetTopToCurrentScrollOffset();
                widget.onPointerUp?.call(
                  context,
                  getInfo(size),
                );
                animateToPosition(widget.calculateTopPosition(
                  getInfo(size),
                ));
              },
              child: (widget.grabbingPosition == WidgetPosition.below ||
                      widget.grabbing == null)
                  ? Column(children: [
                      if (widget.grabbing != null) buildGrabber(),
                      _buildScrollView(),
                      if (widget.bottomWidget != null) widget.bottomWidget!,
                    ])
                  : Stack(
                      children: [
                        Column(children: [
                          if (widget.grabbing != null)
                            Opacity(opacity: 0, child: widget.grabbing),
                          _buildScrollView(),
                          if (widget.bottomWidget != null) widget.bottomWidget!,
                        ]),
                        buildGrabber(),
                      ],
                    ),
            ),
          ),
          if (widget.topPosition == WidgetPosition.above && widgetTop != null)
            widgetTop,
        ],
      );
    });
  }

  CommentSheetInfo getInfo(BoxConstraints size) {
    return CommentSheetInfo(
      velocity: _vt,
      size: size,
      currentTop: fakeTop,
      scrollController: scrollController,
      animatingTarget: _animatingTarget,
      isAnimating: animationController.isAnimating,
    );
  }

  Widget buildGrabber() {
    if (widget.grabbing == null) {
      return const SizedBox();
    }
    return Transform.translate(
      offset: Offset(0, -_scrollOffset),
      child: GestureDetector(
        onVerticalDragUpdate: (detail) {
          setState(() {
            _top = _top + detail.delta.dy;
            widget.onTopChanged?.call(fakeTop);
          });
        },
        onPanCancel: () {
          _movingGrabbing = false;
        },
        onPanEnd: (detail) {
          _movingGrabbing = false;
        },
        onPanDown: (detail) {
          _movingGrabbing = true;
          if (animationController.isAnimating) {
            animationController.stop();
          }
        },
        child: widget.grabbing!,
      ),
    );
  }

  Widget _buildScrollView() {
    return Expanded(
      child: CustomScrollView(
        controller: scrollController,
        physics: widget.scrollPhysics,
        slivers: widget.slivers,
      ),
    );
  }
}

class CommentSheetBouncingScrollPhysics extends ScrollPhysics {
  /// Creates scroll physics that bounce back from the edge.
  const CommentSheetBouncingScrollPhysics({ScrollPhysics? parent})
      : super(parent: parent);

  @override
  CommentSheetBouncingScrollPhysics applyTo(ScrollPhysics? ancestor) {
    return CommentSheetBouncingScrollPhysics(
      parent: buildParent(ancestor),
    );
  }

  /// The multiple applied to overscroll to make it appear that scrolling past
  /// the edge of the scrollable contents is harder than scrolling the list.
  /// This is done by reducing the ratio of the scroll effect output vs the
  /// scroll gesture input.
  ///
  /// This factor starts at 0.52 and progressively becomes harder to overscroll
  /// as more of the area past the edge is dragged in (represented by an increasing
  /// `overscrollFraction` which starts at 0 when there is no overscroll).
  double frictionFactor(double overscrollFraction) =>
      0.52 * math.pow(1 - overscrollFraction, 2);

  @override
  double applyPhysicsToUserOffset(ScrollMetrics position, double offset) {
    assert(offset != 0.0);
    assert(position.minScrollExtent <= position.maxScrollExtent);

    if (!position.outOfRange) {
      return offset;
    }
    if (position.pixels <= 0) {
      return offset;
    }

    final double overscrollPastStart =
        math.max(position.minScrollExtent - position.pixels, 0.0);
    final double overscrollPastEnd =
        math.max(position.pixels - position.maxScrollExtent, 0.0);
    final double overscrollPast =
        math.max(overscrollPastStart, overscrollPastEnd);
    final bool easing = (overscrollPastStart > 0.0 && offset < 0.0) ||
        (overscrollPastEnd > 0.0 && offset > 0.0);

    final double friction = easing
        // Apply less resistance when easing the overscroll vs tensioning.
        ? frictionFactor(
            (overscrollPast - offset.abs()) / position.viewportDimension)
        : frictionFactor(overscrollPast / position.viewportDimension);
    final double direction = offset.sign;

    return direction * _applyFriction(overscrollPast, offset.abs(), friction);
  }

  static double _applyFriction(
      double extentOutside, double absDelta, double gamma) {
    assert(absDelta > 0);
    double total = 0.0;
    if (extentOutside > 0) {
      final double deltaToLimit = extentOutside / gamma;
      if (absDelta < deltaToLimit) {
        return absDelta * gamma;
      }
      total += extentOutside;
      absDelta -= deltaToLimit;
    }
    return total + absDelta;
  }

  @override
  double applyBoundaryConditions(ScrollMetrics position, double value) => 0.0;

  @override
  Simulation? createBallisticSimulation(
      ScrollMetrics position, double velocity) {
    final Tolerance tolerance = this.tolerance;
    if (velocity.abs() >= tolerance.velocity || position.outOfRange) {
      return StopBouncingScrollSimulation(
        spring: spring,
        position: position.pixels,
        velocity: velocity,
        leadingExtent: position.minScrollExtent,
        trailingExtent: position.maxScrollExtent,
        tolerance: tolerance,
      );
    }
    // return null to begin an idle activity
    return null;
  }

  // The ballistic simulation here decelerates more slowly than the one for
  // ClampingScrollPhysics so we require a more deliberate input gesture
  // to trigger a fling.
  @override
  double get minFlingVelocity => kMinFlingVelocity * 2.0;

  // Methodology:
  // 1- Use https://github.com/flutter/platform_tests/tree/master/scroll_overlay to test with
  //    Flutter and platform scroll views superimposed.
  // 3- If the scrollables stopped overlapping at any moment, adjust the desired
  //    output value of this function at that input speed.
  // 4- Feed new input/output set into a power curve fitter. Change function
  //    and repeat from 2.
  // 5- Repeat from 2 with medium and slow flings.
  /// Momentum build-up function that mimics iOS's scroll speed increase with repeated flings.
  ///
  /// The velocity of the last fling is not an important factor. Existing speed
  /// and (related) time since last fling are factors for the velocity transfer
  /// calculations.
  @override
  double carriedMomentum(double existingVelocity) {
    return existingVelocity.sign *
        math.min(0.000816 * math.pow(existingVelocity.abs(), 1.967).toDouble(),
            40000.0);
  }

  // Eyeballed from observation to counter the effect of an unintended scroll
  // from the natural motion of lifting the finger after a scroll.
  @override
  double get dragStartDistanceMotionThreshold => 3.5;
}

class StopBouncingScrollSimulation extends Simulation {
  /// Creates a simulation group for scrolling on iOS, with the given
  /// parameters.
  ///
  /// The position and velocity arguments must use the same units as will be
  /// expected from the [x] and [dx] methods respectively (typically logical
  /// pixels and logical pixels per second respectively).
  ///
  /// The leading and trailing extents must use the unit of length, the same
  /// unit as used for the position argument and as expected from the [x]
  /// method (typically logical pixels).
  ///
  /// The units used with the provided [SpringDescription] must similarly be
  /// consistent with the other arguments. A default set of constants is used
  /// for the `spring` description if it is omitted; these defaults assume
  /// that the unit of length is the logical pixel.
  StopBouncingScrollSimulation({
    required double position,
    required double velocity,
    required this.leadingExtent,
    required this.trailingExtent,
    required this.spring,
    required Tolerance tolerance,
  }) : assert(position != null),
        assert(velocity != null),
        assert(leadingExtent != null),
        assert(trailingExtent != null),
        assert(leadingExtent <= trailingExtent),
        assert(spring != null),
        super(tolerance: tolerance) {
    if (position < leadingExtent) {
      _springSimulation = _underscrollSimulation(position, velocity);
      _springTime = double.negativeInfinity;
    } else if (position > trailingExtent) {
      _springSimulation = _overscrollSimulation(position, velocity);
      _springTime = double.negativeInfinity;
    } else {
      // Taken from UIScrollView.decelerationRate (.normal = 0.998)
      // 0.998^1000 = ~0.135
      _frictionSimulation = FrictionSimulation(0.135, position, velocity);
      final double finalX = _frictionSimulation.finalX;
      if (velocity > 0.0 && finalX > trailingExtent) {
        _springTime = _frictionSimulation.timeAtX(trailingExtent);
        _springSimulation = StopSimulation(
          trailingExtent,
        );
        assert(_springTime.isFinite);
      } else if (velocity < 0.0 && finalX < leadingExtent) {
        _springTime = _frictionSimulation.timeAtX(leadingExtent);
        _springSimulation = StopSimulation(
          leadingExtent,
        );
        assert(_springTime.isFinite);
      } else {
        _springTime = double.infinity;
      }
    }
    assert(_springTime != null);
  }

  /// The maximum velocity that can be transferred from the inertia of a ballistic
  /// scroll into overscroll.
  static const double maxSpringTransferVelocity = 5000.0;

  /// When [x] falls below this value the simulation switches from an internal friction
  /// model to a spring model which causes [x] to "spring" back to [leadingExtent].
  final double leadingExtent;

  /// When [x] exceeds this value the simulation switches from an internal friction
  /// model to a spring model which causes [x] to "spring" back to [trailingExtent].
  final double trailingExtent;

  /// The spring used to return [x] to either [leadingExtent] or [trailingExtent].
  final SpringDescription spring;

  late FrictionSimulation _frictionSimulation;
  late Simulation _springSimulation;
  late double _springTime;
  double _timeOffset = 0.0;

  Simulation _underscrollSimulation(double x, double dx) {
    return ScrollSpringSimulation(spring, x, leadingExtent, dx);
  }

  Simulation _overscrollSimulation(double x, double dx) {
    return ScrollSpringSimulation(spring, x, trailingExtent, dx);
  }

  Simulation _simulation(double time) {
    final Simulation simulation;
    if (time > _springTime) {
      _timeOffset = _springTime.isFinite ? _springTime : 0.0;
      simulation = _springSimulation;
    } else {
      _timeOffset = 0.0;
      simulation = _frictionSimulation;
    }
    return simulation..tolerance = tolerance;
  }

  @override
  double x(double time) => _simulation(time).x(time - _timeOffset);

  @override
  double dx(double time) => _simulation(time).dx(time - _timeOffset);

  @override
  bool isDone(double time) => _simulation(time).isDone(time - _timeOffset);

  @override
  String toString() {
    return '${objectRuntimeType(this, 'BouncingScrollSimulation')}(leadingExtent: $leadingExtent, trailingExtent: $trailingExtent)';
  }
}


class StopSimulation extends Simulation {
  final double stop;

  StopSimulation(this.stop);

  @override
  double dx(double time) {
    return 0;
  }

  @override
  bool isDone(double time) {
    return true;
  }

  @override
  double x(double time) {
    return stop;
  }

}

example/lib/screen_sheet_demo.dart
import 'dart:math';

import 'package:comment_sheet/comment_sheet.dart';
import 'package:flutter/material.dart';

import 'main.dart';

class ScreenSheetDemo extends StatefulWidget {
  const ScreenSheetDemo({Key? key}) : super(key: key);

  @override
  State<ScreenSheetDemo> createState() => _ScreenSheetDemoState();
}

class _ScreenSheetDemoState extends State<ScreenSheetDemo> {
  final scrollController = ScrollController();
  final commentSheetController = CommentSheetController();

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(),
      body: CommentSheet(
        slivers: [
          buildSliverList(),
        ],
        grabbingPosition: WidgetPosition.above,
        initTopPosition: 200,
        calculateTopPosition: calculateTopPosition,
        onTopChanged: (top) {
        },
        scrollController: scrollController,
        grabbing: buildGrabbing(context),
        topWidget: (info) {
          return Positioned(
            top: 0,
            left: 0,
            right: 0,
            height: max(0, info.currentTop),
            child: const Placeholder(
              color: Colors.green,
            ),
          );
        },
        topPosition: WidgetPosition.below,
        bottomWidget: buildBottomWidget(),
        onPointerUp: (
            BuildContext context,
            CommentSheetInfo info,
            ) {
        },
        onAnimationComplete: (
            BuildContext context,
            CommentSheetInfo info,
            ) {
        },
        scrollPhysics: const CommentSheetBouncingScrollPhysics(
          parent: AlwaysScrollableScrollPhysics()),
        commentSheetController: commentSheetController,
        child: const Placeholder(),
        backgroundBuilder: (context) {
          return Container(
            color: const Color(0xFF0F0F0F),
            margin: const EdgeInsets.only(top: 20),
          );
        },
      ),
    );
  }

  double calculateTopPosition(
    CommentSheetInfo info,
  ) {
    final vy = info.velocity.getVelocity().pixelsPerSecond.dy;
    final top = info.currentTop;
    double p0 = 0;
    double p1 = 200;
    double p2 = info.size.maxHeight - 100;

    if (top > p1) {
      if (vy > 0) {
        if (info.isAnimating && info.animatingTarget == p1 && top < p1 + 10) {
          return p1;
        } else {
          return p2;
        }
      } else {
        return p1;
      }
    } else if (top == p1) {
      return p1;
    } else if (top == p0) {
      return p0;
    } else {
      if (vy > 0) {
        if (info.isAnimating && info.animatingTarget == p0 && top < p0 + 10) {
          return p0;
        } else {
          return p1;
        }
      } else {
        return p0;
      }
    }
  }

  Container buildBottomWidget() {
    return Container(
      color: Colors.transparent,
      height: 50,
      child: const Placeholder(
        color: Colors.blue,
      ),
    );
  }

  Widget buildGrabbing(BuildContext context) {
    return const GrabbingWidget();
  }

  Widget buildSliverList() {
    return SliverList(
        delegate: SliverChildBuilderDelegate((context, index) {
      return const ListItemWidget();
    }, childCount: 15));
  }
}


example/lib/main.dart
import 'dart:math';

import 'package:comment_sheet/comment_sheet.dart';
import 'package:example/screen_sheet_demo.dart';
import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({Key? key, required this.title}) : super(key: key);

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  final CommentSheetController commentSheetController =
      CommentSheetController();
  ScrollController scrollController = ScrollController();

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: SafeArea(
        child: Builder(builder: (context) {
          return CustomScrollView(
            physics: const BouncingScrollPhysics(
                parent: AlwaysScrollableScrollPhysics()),
            slivers: [
              LoadingCupertinoSliverRefreshControl(
                onRefresh: () async {
                  await Future.delayed(const Duration(seconds: 3));
                },
              ),
              SliverFillRemaining(
                child: Column(
                  mainAxisAlignment: MainAxisAlignment.center,
                  children: [
                    ListTile(
                      title: const Text("Open Sheet"),
                      onTap: () {
                        showBottomSheet(
                          context: context,
                          backgroundColor: Colors.transparent,
                          builder: (context) {
                            return CommentSheet(
                              slivers: [
                                buildSliverList(),
                              ],
                              grabbingPosition: WidgetPosition.above,
                              initTopPosition: 200,
                              calculateTopPosition: calculateTopPosition,
                              scrollController: scrollController,
                              grabbing: Builder(builder: (context) {
                                return buildGrabbing(context);
                              }),
                              topWidget: (info) {
                                return Positioned(
                                  top: 0,
                                  left: 0,
                                  right: 0,
                                  height: max(0, info.currentTop),
                                  child: const Placeholder(
                                    color: Colors.green,
                                  ),
                                );
                              },
                              topPosition: WidgetPosition.below,
                              bottomWidget: buildBottomWidget(),
                              onPointerUp: (
                                BuildContext context,
                                CommentSheetInfo info,
                              ) {
                                // print("On Pointer Up");
                              },
                              onAnimationComplete: (
                                BuildContext context,
                                CommentSheetInfo info,
                              ) {
                                // print("onAnimationComplete");
                                if (info.currentTop >=
                                    info.size.maxHeight - 100) {
                                  Navigator.of(context).pop();
                                }
                              },
                              commentSheetController: commentSheetController,
                              onTopChanged: (top) {
                                // print("top: $top");
                              },
                              child: const Placeholder(),
                              backgroundBuilder: (context) {
                                return Container(
                                  color: const Color(0xFF0F0F0F),
                                  margin: const EdgeInsets.only(top: 20),
                                );
                              },
                            );
                          },
                        );
                      },
                    ),
                    ListTile(
                      title: const Text("Show Sheet in Stack"),
                      onTap: () {
                        Navigator.push(
                          context,
                          MaterialPageRoute(
                            builder: (context) {
                              return const ScreenSheetDemo();
                            },
                          ),
                        );
                      },
                    ),
                  ],
                ),
              )
            ],
          );
        }),
      ),
    );
  }

  Container buildBottomWidget() {
    return Container(
      color: Colors.transparent,
      height: 50,
      child: const Placeholder(
        color: Colors.blue,
      ),
    );
  }

  double calculateTopPosition(
    CommentSheetInfo info,
  ) {
    final vy = info.velocity.getVelocity().pixelsPerSecond.dy;
    final top = info.currentTop;
    double p0 = 0;
    double p1 = 200;
    double p2 = info.size.maxHeight - 100;

    if (top > p1) {
      if (vy > 0) {
        if (info.isAnimating && info.animatingTarget == p1 && top < p1 + 10) {
          return p1;
        } else {
          return p2;
        }
      } else {
        return p1;
      }
    } else if (top == p1) {
      return p1;
    } else if (top == p0) {
      return p0;
    } else {
      if (vy > 0) {
        if (info.isAnimating && info.animatingTarget == p0 && top < p0 + 10) {
          return p0;
        } else {
          return p1;
        }
      } else {
        return p0;
      }
    }
  }

  Widget buildGrabbing(BuildContext context) {
    return const GrabbingWidget();
  }

  Widget buildSliverList() {
    return SliverList(
      delegate: SliverChildBuilderDelegate((context, index) {
        return const ListItemWidget();
      }, childCount: 20),
    );
  }
}

class GrabbingWidget extends StatelessWidget {
  const GrabbingWidget({
    Key? key,
  }) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Container(
      decoration: const BoxDecoration(
        color: Color(0xFF0F0F0F),
        borderRadius: BorderRadius.only(
          topLeft: Radius.circular(12),
          topRight: Radius.circular(12),
        ),
      ),
      padding: const EdgeInsets.only(top: 10),
      child: Column(
        mainAxisSize: MainAxisSize.min,
        children: [
          Container(
            width: 40,
            height: 4,
            margin: const EdgeInsets.only(top: 4),
            decoration: BoxDecoration(
                color: Colors.white60,
                borderRadius: BorderRadius.circular(100)),
          ),
          Row(
            children: [
              const Padding(
                padding: EdgeInsets.only(left: 15),
                child: Text(
                  "Bình luận",
                  style: TextStyle(
                    fontWeight: FontWeight.w700,
                    color: Colors.white,
                    fontSize: 16,
                  ),
                ),
              ),
              const Padding(
                padding: EdgeInsets.only(left: 6.0),
                child: Text(
                  "48",
                  style: TextStyle(
                    fontWeight: FontWeight.w400,
                    color: Colors.white70,
                    fontSize: 13,
                  ),
                ),
              ),
              const Spacer(),
              const Icon(
                Icons.menu_sharp,
                size: 26,
                color: Colors.white,
              ),
              Padding(
                padding: const EdgeInsets.only(left: 10.0, right: 10),
                child: IconButton(
                  onPressed: () {
                    Navigator.of(context).pop();
                  },
                  icon: const Icon(
                    Icons.close,
                    size: 26,
                    color: Colors.white,
                  ),
                ),
              )
            ],
          ),
          Container(
            color: const Color(0xFF292929),
            width: double.infinity,
            height: 1,
          ),
        ],
      ),
    );
  }
}

class ListItemWidget extends StatelessWidget {
  const ListItemWidget({
    Key? key,
  }) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Material(
      color: const Color(0xFF0F0F0F),
      child: InkWell(
        onTap: () {
          Navigator.of(context).pop();
        },
        child: Container(
          width: double.infinity,
          padding:
              const EdgeInsets.only(top: 12, bottom: 0, left: 10, right: 10),
          child: Row(
            crossAxisAlignment: CrossAxisAlignment.start,
            children: [
              Padding(
                padding: const EdgeInsets.only(left: 5, right: 10),
                child: ClipOval(
                        child: Image.network(
                          "https://yt3.ggpht.com/yti/AJo0G0kUnHqoybmWPJG4GNm0G-lfCiCPbEP62v5tq9PZsA=s48-c-k-c0x00ffffff-no-rj",
                          width: 25,
                          height: 25,
                        ),
                      ),
              ),
              Expanded(
                child: Column(
                  crossAxisAlignment: CrossAxisAlignment.start,
                  children: [
                    const Text(
                      'Andrea Quintanilla * 3 tháng trước',
                      style: TextStyle(
                          fontSize: 10,
                          fontWeight: FontWeight.w400,
                          color: Color(0xFFAEAEAE)),
                    ),
                    const Padding(
                      padding: EdgeInsets.only(top: 6, bottom: 12),
                      child: Text(
                        'Que buen trabajo, que buenos enganches, genial!!!!  MTV la tenes adentro, jajaja. Saludos cordiales desde Buenos Aires, Argentina, Argentina, Argentina!',
                        style: TextStyle(
                            fontSize: 13,
                            fontWeight: FontWeight.w400,
                            color: Color(0xFFF6F6F6)),
                      ),
                    ),
                    Row(
                      children: const [
                        Icon(
                          Icons.thumb_up_outlined,
                          size: 15,
                          color: Colors.white,
                        ),
                        Padding(
                          padding: EdgeInsets.only(left: 16.0, right: 16.0),
                          child: Icon(
                            Icons.thumb_down_alt_outlined,
                            size: 15,
                            color: Colors.white,
                          ),
                        ),
                        Icon(
                          Icons.comment_outlined,
                          size: 15,
                          color: Colors.white,
                        ),
                      ],
                    )
                  ],
                ),
              )
            ],
          ),
        ),
      ),
    );
  }
}
